pkg/pipeline/encode/encode_s3.go

```diff
@@ -0,0 +1,219 @@
+	pendingEntries    []config.GenericMap
+	mutex             *sync.Mutex
+	expiryTime        int64
+	exitChan          <-chan struct{}
+	streamId          string
+	intervalStartTime time.Time
+	sequenceNumber    int64
+}
+
+type s3WriteEntries interface {
+	putObject(bucket string, objectName string, object map[string]interface{}) error
+}
+
+type encodeS3Writer struct {
+	s3Client *minio.Client
+	s3Params *api.EncodeS3
+}
+
+func (s *encodeS3) writeObject() error {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	nLogs := len(s.pendingEntries)
+	if nLogs > s.s3Params.BatchSize {
+		nLogs = s.s3Params.BatchSize
+	}
+	now := time.Now()
+	object := s.GenerateStoreHeader(s.pendingEntries[0:nLogs], s.intervalStartTime, now)
+	year := fmt.Sprintf("%04d", now.Year())
+	month := fmt.Sprintf("%02d", now.Month())
+	day := fmt.Sprintf("%02d", now.Day())
+	hour := fmt.Sprintf("%02d", now.Hour())
+	seq := fmt.Sprintf("%08d", s.sequenceNumber)
+	objectName := s.s3Params.Account + "/year=" + year + "/month=" + month + "/day=" + day + "/hour=" + hour + "/stream-id=" + s.streamId + "/" + seq
+	log.Debugf("S3 writeObject: objectName = %s", objectName)
+	log.Debugf("S3 writeObject: object = %v", object)
+	s.pendingEntries = s.pendingEntries[nLogs:]
+	s.intervalStartTime = now
+	s.expiryTime = now.Unix() + s.s3Params.WriteTimeout
+	s.sequenceNumber++
+	// send object to object store
+	err := s.s3Writer.putObject(s.s3Params.Bucket, objectName, object)
+	if err != nil {
+		log.Errorf("error in writing object: %v", err)
+	}
+	return err
+}
+
+func (s *encodeS3) GenerateStoreHeader(flows []config.GenericMap, startTime time.Time, endTime time.Time) map[string]interface{} {
+	object := make(map[string]interface{})
+	// copy user defined keys from config to object header
+	for key, value := range s.s3Params.ObjectHeaderParameters {
+		object[key] = value
+	}
+	object["version"] = flpS3Version
+	object["capture_start_time"] = startTime.Format(time.RFC3339)
+	object["capture_end_time"] = endTime.Format(time.RFC3339)
+	object["number_of_flow_logs"] = len(flows)
+	object["flow_logs"] = flows
+
+	return object
+}
+
+func (s *encodeS3) createObjectTimeoutLoop() {
+	log.Debugf("entering createObjectTimeoutLoop \n")
+	ticker := time.NewTicker(time.Duration(s.s3Params.WriteTimeout) * time.Second)
+	for {
+		select {
+		case <-s.exitChan:
+			log.Debugf("exiting createObjectTimeoutLoop because of signal")
+			return
+		case <-ticker.C:
+			log.Debugf("createObjectTimeoutLoop timer expire\n")
+			now := time.Now().Unix()
+			log.Debugf("time now = %d, expiryTime = %d \n", now, s.expiryTime)
+			if now >= s.expiryTime {
+				_ = s.writeObject()
+			}
+		}
+	}
+}
+
+// Encode queues entries to be sent to object store
+func (s *encodeS3) Encode(entry config.GenericMap) {
+	log.Debugf("Encode S3, entry = %v", entry)
+	s.mutex.Lock()
+	s.pendingEntries = append(s.pendingEntries, entry)
+	s.mutex.Unlock()
+	s.recordsWritten.Inc()
+	if len(s.pendingEntries) >= s.s3Params.BatchSize {
```

Comment:
This access to `s.pendingEntries` is not guarded by the mutex.
If there is a context switch to the TimeoutLoop right after a truthy evaluation of this condition but before the call to writeObject(), then we might end up with writing empty objects.